packages/plugin-npm/sources/npmHttpUtils.ts

```diff
@@ -64,15 +66,166 @@ export function getIdentUrl(ident: Ident) {
+// - an on-disk cache, for exact version matches and to avoid refetching the metadata if the resource hasn't changed on the server
+
+const PACKAGE_METADATA_CACHE = new Map<IdentHash, PackageMetadata>();
+
+/**
+ * Caches and returns the package metadata for the given ident.
+ *
+ * Note: This function only caches and returns specific fields from the metadata.
+ * If you need other fields, use the uncached {@link get} or consider whether it would make more sense to extract
+ * the fields from the on-disk packages using the linkers or from the fetch results using the fetchers.
+ */
+export async function getPackageMetadata(ident: Ident, {project, registry, headers, version, ...rest}: GetPackageMetadataOptions): Promise<PackageMetadata> {
+  const {configuration} = project;
+
+  const cachedInMemory = PACKAGE_METADATA_CACHE.get(ident.identHash);
+  if (cachedInMemory)
+    return cachedInMemory;
+
+  registry = normalizeRegistry(configuration, {ident, registry});
+
+  const registryFolder = getRegistryFolder(configuration, registry);
+  const identPath = ppath.join(registryFolder, `${structUtils.slugifyIdent(ident)}.json`);
+
+  let cachedOnDisk: CachedMetadata | null = null;
+
+  // We bypass the on-disk cache for security reasons if the lockfile needs to be refreshed,
+  // since most likely the user is trying to validate the metadata using hardened mode.
+  if (!project.lockfileNeedsRefresh) {
+    try {
+      cachedOnDisk = await xfs.readJsonPromise(identPath) as CachedMetadata;
+
+      if (typeof version !== `undefined` && typeof cachedOnDisk.metadata.versions[version] !== `undefined`) {
+        return cachedOnDisk.metadata;
+      }
+    } catch {}
+  }
+
+  return await get(getIdentUrl(ident), {
+    ...rest,
+    customErrorMessage: customPackageError,
+    configuration,
+    registry,
+    ident,
+    headers: {
+      ...headers,
+      // We set both headers in case a registry doesn't support ETags
+      [`If-None-Match`]: cachedOnDisk?.etag,
+      [`If-Modified-Since`]: cachedOnDisk?.lastModified,
+    },
+    wrapNetworkRequest: async executor => async () => {
+      const response = await executor();
+
+      if (response.statusCode === 304) {
+        if (cachedOnDisk === null)
+          throw new Error(`Assertion failed: cachedMetadata should not be null`);
+
+        return {
+          ...response,
+          body: cachedOnDisk.metadata,
+        };
+      }
+
+      const packageMetadata = pickPackageMetadata(JSON.parse(response.body.toString()));
+
+      PACKAGE_METADATA_CACHE.set(ident.identHash, packageMetadata);
+
+      const metadata: CachedMetadata = {
+        metadata: packageMetadata,
+        etag: response.headers.etag,
+        lastModified: response.headers[`last-modified`],
+      };
+
+      await xfs.mkdirPromise(registryFolder, {recursive: true});
+      await xfs.writeJsonPromise(identPath, metadata, {compact: true});
```

Comment:
This write needs to be atomic.

Reply:
Good idea, I've made the write atomic and also changed the in-memory cache to cache the entire promise to ensure that the disk and network are only hit once per process for a given ident (could have led to multiple calls writing to the same file at the same time).